### DISCUSSION TITLE
Proper macOS menus compatible with Gtk3 and future proof for Gtk4

### DISCUSSION THREAD

**Discussion by DominusExult** [jralls/gtk-osx-build#93] [Q&A]
We are slowly porting our Gtk3 project Exult Studio from Gtkmenu to Gmenu and are preparing to eventually port to Gtk4 (mostly waiting for Drag and drop on macOS to work properly - last time it still had problems with the demo).
So far we do not have "proper" macOS menus, just the menu bar of the application, not using the macOS menu bar.
Looking into it we hit some roadblocks:
- Everything points at https://gitlab.gnome.org/GNOME/gtk-mac-integration
- but the docs at https://jralls.github.io/gtk-mac-integration/GtkosxApplication.html say that this is 

> This library is obsolescent in Gtk3 as there are built-in features in GLib and Gtk to integrate Gtk applications with macOS.
- trying to find documentation on these built in features in https://docs.gtk.org/gtk3/macos.html leads me to https://docs.gtk.org/gtk3/macos.html -> https://wiki.gnome.org/Projects/GTK/OSX/Integration 

> Since macOS integration is available directly in Gtk using GApplication

And then I'm reading that there won't be a normal menu anymore in Gtk4? I'm very confused :)
Trying to find examples for Gtk3 applications on macOS *without* using gtk-mac-integration proved futile. Is there any example?

Edit: disregard mostly. Inkscape is another example besides Gimp 2.99 preview of a Gtk3 app without gtk-mac-integration but proper macOS menu.
On the Gtk4 side, I found that the demo "Application Class" also shows the proper menu on macOS.

**Comment by jralls** [comment]
If you've based your app on [GApplication](https://docs.gtk.org/gio/class.Application.html)  as [recommended by the migration guide](https://docs.gtk.org/gtk4/migrating-3to4.html#stop-using-gtk_main-and-related-apis) and [GMenuModel](https://docs.gtk.org/gio/class.MenuModel.html) it should Just Work.

**Comment by DominusExult** [comment]
Thanks, we are getting there now. Properly showing the menus now (though no item is yet working for some reason.

A mistake I noticed in gtk3-demo-application and gtk4-demo-application:
- gtk3 menu items that should be in a "File" menu are under the application menu
- gtk4 those menu items are still in the application menu, but there is a "File" menu now, which is now showing the items that were previously under the "Preferences" menu. I guess something went wrong when someone wanted to move the application items under the "File" menu.

I filed an issue for that at https://gitlab.gnome.org/GNOME/gtk/-/issues/6524

**Reply by jralls** [reply]
I see what you mean. I haven't migrated either of the apps I help maintain to GApplication so I don't know if it will require fixing in Gtk or if it's just that the demos didn't do the layout right.

**Reply by DominusExult** [reply]
As the "Application" menu of our Gtk3 app reverted to the system defaults once we had the menus properly besides it, my guess is that the gtk3-application-demo just has the layout wrong. 
And then someone messed it up further for gtk4-application-demo (probably by making a mistake when adding the "Files" menu).

**Reply by DominusExult** [reply]
We have a preliminary port of Gtk4 working and it shows the same behavior as the gtk-demo-application -> the "application" menu in the macOS menubar is not showing the macOS menu but the first submenu defined in the menu xml.
Seems this is a bug in Gtk4. As I understand gtk-shell-shows-menubar is supposed to take over the job of gtk-shell-shows-app-menu which was handling that part in Gtk3.
(If I toggle gtk-shell-shows-menubar in the debug app it shows the menu bar correctly in our app, but toggling gtk-shell-shows-app-menu does nothing there. In gtk3 it added a menu item "Application" to the in-app menu when disabled)

**Reply by jralls** [reply]
That does seem to be a regression. Please open an issue at https://gitlab.gnome.org/GNOME/gtk/~/issues. Please be sure to set the macOS and GTK4 tags on it.

**Reply by DominusExult** [reply]
Yes, I think it's a regression in gtkapplications-quartz.c. The goal was to make sure to always show a menu but accidentally nuked the Application menu when there is a menubar. So now the macOS Application menu only shows when there is NO menubar.
Instead of if (menubar)… else (app_menu code) it should have left the else away, so the app_menu code is always applied. So
if(menubar)
  [codeline]
App_menu code

I added a fix for gtk3-demo-application in the issue https://gitlab.gnome.org/GNOME/gtk/-/issues/6524 and once I figured out how to fix the gtk4 bug I'll open a new issue and refer to it here and there.

**Reply by DominusExult** [reply]
After spending way too much time on this: it's complicated :)
See https://gitlab.gnome.org/GNOME/gtk/-/issues/6539

**Reply by DominusExult** [reply]
With a lot of help I've put up a good patch. But it's really annoying as it basically reinvents the wheel, adding back an implementation for a custom Application menu that was way more elegantly done with `gtk_application_set_app_menu`.
Was there any reason for this other than forcing people on some new way of doing menus that may not even be popular? I'm honestly interested for the reason.

**Reply by jralls** [reply]
I don't know, I haven't followed closely Gtk4 development. I suggest the [Gnome Discourse](https://discourse.gnome.org/c/platform/5) or the #gtk [IRC channel](https://wiki.gnome.org/GettingInTouch/IRC).

**Comment by DominusExult** [comment]
So we've successfully ported from Gtkmenu to Gmenu on Gtk3 with proper macOS  menubar menus.
Things that I wish I would have realized sooner and want to record here in case anyone ever stumbles over it:
- There are hardcoded action names that Gtk will use to populate the standard macOS application menu (menu with the name of the app to the right of the apple symbol): app.about, app.preferences, app.quit
- Gtk will use these for the application menu even if they are not menu items in your menu xml file.
- If they are in the menu xml file and you want to prevent them showing up twice you can add this attribute to the item `<attribute name="hidden-when">macos-menubar</attribute>`
- If they are in the menu xml file and you declare a keyboard shortcut (e.g. `<attribute name="accel">&lt;Control&gt;q</attribute>`) this keyboard shortcut will take precedence over the system's default keyboard shortcut and even show in the application menu. Regardless whether you added the `hidden-when` attribute.
- for keyboard shortcuts that should be `Control + key` on linux/Windows but `Command + key` on macOS use `Primary` which will use the correct one, depending on the OS (e.g. `<attribute name="accel">&lt;Primary&gt;q</attribute>`)

These are not criticism of the way it is handled, I just wasn't able to find that information easily (or didn't know where to look - I'm sure it's somewhere and I'll slap my face when I see it) and ended up piecing the puzzle together by looking at the menu xmls of Gimp and Inkscape. The hardcoded action names surprised me, though :)


> though no item is yet working for some reason

that was fixed by switching from app_group to application

![image](https://github.com/jralls/gtk-osx-build/assets/3313896/8de51ad3-d4af-474e-96a2-a5de2e042256)

**Reply by jralls** [reply]
Thanks. Maybe you could write a PR to add that info to the GMenuModel documentation somewhere.in the [GMenuModel documentation](https://gitlab.gnome.org/GNOME/glib/-/blame/main/gio/gmenumodel.c?ref_type=heads#L30). That would be easier to find than remembering that there's a nugget buried in a comment here.

**Reply by DominusExult** [reply]
Once we successfully ported to Gtk4 I'll see where to add it. Maybe the wiki is a better place 🤷‍♂️
I was amazed that I couldn't find these information anywhere but in other project's code even by now knowing what to search for.

**Reply by rolandlo** [reply]
What's the equivalent of  [`gtkosx_application_set_help_menu`](https://github.com/GNOME/gtk-mac-integration/blob/master/src/gtkosxapplication_quartz.c#L964-L997) for a `GApplication` using `GMenu`s? In our app the spotlight search entry only shows up when the language is English (and the menu name "_Help" is untranslated), it doesn't show up with other languages.

**Reply by jralls** [reply]
@rolandlo I don't know. @DominusExult have you figured that out?

**Reply by DominusExult** [reply]
I will take a look next week, I think I stumbled over that

**Reply by rolandlo** [reply]
@DominusExult Sorry to bother you again, but did you look into the question on setting the help menu? If so, what is the solution?

**Reply by DominusExult** [reply]
Sorry, I forgot and just now tried to figure it out but I found I cannot. Maybe it's something to discuss upstream.

**Reply by jralls** [reply]
Yeah, there's not much we can do about it here. I don't see any existing [gtk issues](https://gitlab.gnome.org/GNOME/gtk/-/issues/?sort=created_date&state=all&label_name%5B%5D=GTK4&label_name%5B%5D=MacOS&first_page_size=20) addressing it so please open one.
